### ISSUE TITLE
2021/09/14 - 2021/09/27

### ISSUE THREAD

**Issue by njs03332** [njs03332/nlp_study#1]
Pytorch로 배우는 자연어 처리 1장, 2장 공부해오기
중요한 부분만 각자 정리해서 코멘트로 추가하기

**Comment by njs03332** [comment]
## 1장. 소개
### 1.1 지도학습
* 샘플 = 입력 = x
* 타깃 = 레이블 = y
* 모델: 샘플 x를 받아 타깃 레이블값을 예측하는 함수
* 파라미터 = 가중치 = w
* 예측: 모델이 추측하는 타깃값 y hat
* 손실 함수: 훈련 데이터에 대한 예측이 타깃과 얼마나 멀리 떨어져있는지 비교하는 함수 L
  * 손실: 실수 스칼라값, 손실이 낮을수록 타깃 예측을 더 잘하는 모델
* 지도학습: 샘플 n개 전체의 누적 손실을 최소화하는 최적의 파라미터를 찾는 과정
#### 확률적 경사 하강법
* (배치) 경사 하강법 (batch gradient descent): 주어진 데이터셋에서 손실 함수를 최소화하는 파라미터값 (근)의 초기값을 추측한 후, 목적 함수 (objective function)의 값이 임계점 아래로 내려갈 때까지 파라미터 반복 업데이트 (전통적인 경사 하강법)
* 확률적 (stochastic) 경사 하강법: data point를 일부 랜덤하게 선택하여 gradient를 계산하는 방법
  * 순수 SGD: 데이터 포인트 하나 사용 (거의 사용 x)
  * 미니배치 SGD: 여러개 사용
  * 일반적인 SGD 알고리즘의 수렴 속도를 높이는 파생 알고리즘들이 개발됨
* 역전파 (backpropagation): 파라미터를 반복적으로 업데이트하는 과정
  * 정방향 계산 (forward pass): 현재 파라미터값으로 입력을 평가하여 손실 함수를 계산
  * 역방향 계산 (backward pass): 손실의 그레이디언트를 사용하여 파라미터를 업데이트

### 1.2 샘플과 타깃의 인코딩
* 입력 인코딩: 샘플과 타깃을 벡터나 텐서의 수치로 표현
#### 1.2.1 원-핫 인코딩
* 휴리스틱 기반의 다눈한 카운트 방식 표현
* 0 벡터에서 시작해 문장이나 문서에 등장하는 단어에 상응하는 원소를 1로 설정
* 구, 문장, 문서의 원-핫 벡터 = 구성하는 단어의 원-핫 표현을 단순히 논리합
#### 1.2.2 TF (Term-Frequency) 표현
* 구, 문장, 문서의 TF 표현은 소속 단어의 원-핫 표현을 합해 만들어짐
  * 각 원소는 해당 단어가 문장에 등장하는 횟수
* BoW (Bag of Words)라고도 부름

#### 1.2.3 TF-IDF (Term-Frequency-Inverse-Document-Frequency) 표현
* 역문서 빈도 (IDF): 벡터 표현에서 흔한 토큰의 점수를 낮추고 드문 토큰의 점수를 높임
IDF(w) = log(전체 문서 개수 / 단어 w를 포함한 문서 개수)
* TF-IDF 점수 = TF(w) * IDF(w)
* 딥러닝에서는 주로 TF-IDF와 같은 경험적인 방법 대신 정수 인덱스를 사용한 원-핫 인코딩과 특수한 임베딩 룩업 (embedding lookup' 층으로 신경망의 입력을 구성함

#### 1.2.4 타깃 인코딩
* 타깃 변수가 텍스트 형태일 때: 원-핫 인코딩과 같은 방식 사용 (예: 기계 번역, 요약, 질의응답 문제)
* 타깃 변수가 범주형일 때: 레이블마다 고유한 인덱스를 부여하는 방식 사용
  * 출력 레이블 수가 커지면 문제가 될 수 있음 (예: 언어 모델링)
* 타깃 변수가 수치일 때: 여러 방법이 있음
  * 타깃을 범주형 구간 (bin)으로 바꾸고, 순서가 있는 분류 문제로 다루는 방법
  * 구간은 균등하지 않아도 되며, 데이터에 따라 나눌 수 있음

### 1. 3 계산 그래프
* 계산 그래프를 사용하면 지도학습 시스템에서 데이터 흐름을 간편하게 구현할 수 있음
* 계산 그래프는 수학식을 추상적으로 모델링한 것
  * 딥러닝에서는 Theano, TensorFlow, PyTorch와 같은 계산 그래프의 구현이 부가적으로 _자동 미분_을 구현함
* 계산 그래프의 표현
  * 노드: 수학 연산을 나타냄
  * 에지: 노드로 들어가는 에지 = 입력, 나가는 에지 = 출력

### 1.4 파이토치 기초
* 파이토치는 다른 프레임워크와 달리 _테이프 기반_ 자동미분 방식을 구현함
  * 계산 그래프를 동적으로 정의하고 실행할 수 있음
  * 디버깅이 편리, 복잡한 모델을 손쉽게 만들 수 있음
* 동적 계산 그래프 vs. 정적 계산 그래프
  * 정적: 먼저 계산 그래프를 정의하고 컴파일한 후 실행 - 제품 시스템과 모바일 환경에서 효율적
  * 동적: 실행할 때마다 모델을 컴파일하지 않아도 도미 - 연구나 개발할 때 편리
    * 특히 입력마다 다른 그래프 구조가 만들어질 수 있는 NLP 모델링 작업에 적합
* 텐서: 다차원 데이터를 담은 수학 객체
  * 0차: 하나의 스칼라
  * 1차: 숫자 배열 또는 벡터
  * 2차: 행렬
  * n차: n차원 스칼라 배열

**Comment by danbi5228** [comment]
* 딥러닝의 목적은 표현학습이므로, TF-IDF와 같이 경험적인 방법으로 입력을 인코딩하지 않음
* 주로 정수 인덱스를 사용한 원-핫 인코딩과 특수한 임베딩 룩업 층으로 신경망의 입력을 구성함

## 1.2.2 TF 표현
* TF : 문서(단어) 빈도 term frequency
* 단순히 소속 단어의 원-핫 표현을 합해 만들 수 있음. 즉, 등장 횟수에 비례하여 단어에 가중치 부여
* ex) vocabulary = {time, fruit, flies, like, a, an, arrow, banana}
   * string1 : Fruit flies like time files a fruit
   * string1의 TF 표현: [1, 2, 2, 1, 1, 0, 0, 0]
   * string1의 단어 time의 TF, TF(time): 1 

## 1.2.3 TF-IDF 표현
* IDF : 역문서 빈도 Inverse Document Frequency
* 흔한 토큰의 점수를 낮추고 드문 토큰의 점수를 높임
* TF-IDF: 단어 빈도와 역문서빈도를 사용하여 단어의 중요도를 계산. 값이 높을수록 중요한 단어로 판단
* 토근 w 의 IDF, IDF(w) = log(N/n<sub>w</sub>)
   * N: 전체 문서 개수
   * n<sub>w</sub>: 단어 w를 포함한 문서 개수
* sklearn의 경우 분모가 0이 되는 상황을 방지하고자 분모,분자에 +1을 더하고, \
모든 문서에 포함된 단어가 있을 때 IDF가 0이 되지 않도록 마지막에 +1을 해줌
   * log( (N+1) / (n<sub>w</sub>+1) )+1
* ex) vocabulary = {time, fruit, flies, like, a, an, arrow, banana}, 사이킷런 기준
   * string1: Time flies like an arrow
   * string2: Fruit flies like an banana
   * string1의 토큰 time의 TF-IDF 계산
      * TF: 1
      * IDF: log( (2+1) / (1+1) ) + 1
      * TF*IDF: 1.405465 ...
      * L2 norm 정규화(사이킷런 추가작업)를 거친 최종 tf-idf 값 : 0.4992
* 필요에 따라 직접 구현하거나 위 예시 내용처럼 sklearn을 이용할 수 있음
   * 참고페이지 추가 : https://wikidocs.net/31698

## Pytorch 라이브러리 사용하기
* 1.4.5 인덱싱, 슬라이싱, 연결
   * numpy 와 비슷한 구조
   * 참고: https://nittaku.tistory.com/107
---
* 공부의 목적이 딥러닝~자연어처리 이기는 하지만, 일단 NLP에서 역사가 깊은 기법이라 짚고 넘어가야 할 것 같아서 한번 더 언급
* 개인적으로 numpy 구조에 익숙해지는데 시간이 필요할 것으로 예상됨ㅎ..
* 예시 내용은 모두 책에 있음
   * P.24 ~ P.27 참고
---
(추후 삭제예정)
* 3장 전체
* 3장 \~77 + 1장 연습문제 (이후 78\~3장 끝까지? 흠..)

**Comment by ssm0318** [comment]
![computational_graph](https://user-images.githubusercontent.com/25769209/135270379-b1298b80-7042-42a5-a384-41af67a34f06.png)
